grails-app/services/org/pih/warehouse/data/MigrationService.groovy

```diff
@@ -1197,4 +1205,84 @@ class MigrationService {
+
+
+    /**
+     * Create missing transaction sources for inventory import transactions that don't have one.
+     *
+     * @param location The location for which to create missing transaction sources.
+     * @return A map with the number of inventory counts processed and the number of transaction sources created.
+     */
+    Map<String, Integer> createMissingInventoryImportTransactionSources(Location location) {
+        List<InventoryCount> inventoryCounts = getInventoryImportTransactionsWithMissingTransactionSource(location)
+
+        // We base the candidates on the inventory count view as it has the transactions (e.g. baseline + adjustment pair) associated as one record
+        Set<List<Transaction>> transactionPairs = inventoryCounts
+                // We group inventory count records by transaction and associated transaction (if both exists (baseline+adjustment), associated transaction is the adjustment)
+                .groupBy { [it.transaction, it.associatedTransaction] }
+                .keySet()
+
+        int transactionSourcesCreated = 0
+        transactionPairs.each { List<Transaction> transactionPair ->
+            // If both exists, the associated transaction is an adjustment, and the mainTransaction is baseline
+            // If only main transaction exists, this might be either baseline or adjustment transaction alone
+            Transaction mainTransaction = transactionPair[0]
+            Transaction associatedTransaction = transactionPair[1]
+
+            // There might be a case in import, where we might have a few inventory count records for the same transaction.
+            // This might be the case when there are multiple products in the import file for the same transaction
+            // and for one of them we would create only a baseline, for another one only adjustment, and for another both baseline + adjustment
+            // So if while looping we identify that for one transaction we already have the source, use it, as it represents the same "system action".
+            TransactionSource existingTransactionSource = mainTransaction.transactionSource ?: associatedTransaction?.transactionSource
+            if (existingTransactionSource) {
+                mainTransaction.transactionSource = existingTransactionSource
+                associatedTransaction?.transactionSource = existingTransactionSource
+                return
+            }
+            TransactionSource transactionSource =
+                    inventoryImportProductInventoryTransactionService.createMissingInventoryImportTransactionSource(AuthService.currentLocation)
+            mainTransaction.transactionSource = transactionSource
+            if (associatedTransaction) {
+                associatedTransaction.transactionSource = transactionSource
+            }
+            transactionSourcesCreated++
+        }
+        return [inventoryCounts: inventoryCounts.size(), transactionSourcesCreated: transactionSourcesCreated]
+    }
```

Comment:
nitpick: since this class is getting really huge, you could consider putting all this logic into a TransactionSourceMigrationService class